File: problems/combine-two-strings/combine-two-strings.java

public class Shuffle {

	public static void main(String[] args) {
		System.out.println(is_shuffle("abc", "def", "dabecf"));
		System.out.println(is_shuffle("bac", "def", "dabecf"));
		System.out.println(is_shuffle("otto", "anna", ""));
		System.out.println(is_shuffle("otto", "anna", "oatntnoa"));
	}

	private static boolean is_shuffle(String string1, String string2, String string3) {
		if (string1.length() + string2.length() != string3.length()) {
			return false;
		}
		
		char[] str1 = string1.toCharArray();
		char[] str2 = string2.toCharArray();
		int ptr1 = 0, ptr2 = 0;
		for (char c : string3.toCharArray()) {
			if (ptr1 < string1.length() && c == str1[ptr1]) {
				ptr1++;
				continue;
			} else if (ptr2 < string2.length() && c == str2[ptr2]) {
				ptr2++;
				continue;
			} else {
				return false;
			}
		}
		return true;
		
	}
}

File: problems/combine-two-strings/combine-two-strings.php
<?php

declare(strict_types=1);

function is_shuffle(string $string1, string $string2, string $string3) : bool {
    if (strlen($string1) + strlen($string2) !== strlen($string3)) {
        return false;
    }

    $ptr1 = 0;
    $ptr2 = 0;
    foreach (str_split($string3) as $c) {
        if ($ptr1 < strlen($string1) && $c === $string1[$ptr1]) {
            $ptr1++;
            continue;
        } else if ($ptr2 < strlen($string2) && $c == $string2[$ptr2]) {
            $ptr2++;
            continue;
        } else {
            return false;
        }
    }
    return true;
}

assert(is_shuffle('abc', 'def', 'dabecf') === true);
assert(is_shuffle('bac', 'def', 'dabecf') === false);
assert(is_shuffle('otto', 'anna', '') === false);
assert(is_shuffle('otto', 'anna', 'oatntnoa') === true);
